src/libpgmoneta/io.c

```diff
@@ -0,0 +1,104 @@
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * 3. Neither the name of the copyright holder nor the names of its contributors
+ * may be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "lz4_compression.h"
+#include <io.h>
+#include <logging.h>
+
+#include <stdint.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+
+ssize_t
+pgmoneta_write_file(int fd, void* buffer, size_t bytes)
+{
+   ssize_t bytes_written = 0;
+   int flags = fcntl(fd, F_GETFL);
+   int misalignment = (uintptr_t)buffer % BLOCK_SIZE;
+   int odd_bytes = bytes % BLOCK_SIZE;
+
+   if (flags & O_DIRECT)
+   {
+      if (misalignment != 0)
+      {
+         void* aligned_buf =
+            pgmoneta_aligned_malloc(((bytes) / BLOCK_SIZE + 1) * BLOCK_SIZE);
+         memcpy(aligned_buf, buffer, bytes);
+         buffer = aligned_buf;
+      }
+
+      if (odd_bytes != 0)
+      {
+         char* tmp_buf = (char*)buffer;
+         while (bytes_written + 8192 < bytes)
+         {
+            bytes_written += write(fd, tmp_buf + bytes_written, 8192);
+         }
+         if (fcntl(fd, F_SETFL, flags & ~O_DIRECT))
+         {
+            pgmoneta_log_error("failed to disable O_DIRECT");
+            return -1;
+         }
+
+         bytes_written += write(fd, tmp_buf + bytes_written, bytes - bytes_written);
+
+         if (fcntl(fd, F_SETFL, flags))
+         {
+            pgmoneta_log_error("failed to re-enable O_DIRECT");
+            return -1;
+         }
+      }
+      else
+      {
+         bytes_written = write(fd, buffer, bytes);
+      }
+
+   }
+   else
+   {
+      bytes_written = write(fd, buffer, bytes);
+   }
+
+   if (bytes_written < 0)
+   {
+      pgmoneta_log_debug("error writing to fd: %d", fd);
```

Comment:
Should be an error, and error -> Error.

You should include the the error string, and then clear errno